### PR TITLE
Fix realtime entropy patching

### PR DESCRIPTION
**Pull request by Vectorrent**
This is a followup to the comment [made here](https://github.com/facebookresearch/blt/pull/25#issuecomment-2599152444).

I want to load and train an entropy model alongside of the latent model - rather than loading it from checkpoint. The current code does not support this, so I added an additional `PatcherArgs` attribute, called `entropy_model`. Here, you can pass a Pytorch module directly, instead of a checkpoint path.

I also removed a `self.output_proj` variable, because it crashes when using entropy patching. The `entropy_model_checkpoint_dir` attribute does not exist on `LocalModelArgs`. Regardless, this variable is unused - and probably not necessary?

Finally, I removed a logger warning that was extremely spammy. We are already gating this warning with the `BLT_SUPPRESS_ATTN_ERROR` environment variable, and I'm assuming you didn't intend on throwing the same warning with every forward pass. I'm sure this was overlooked, and you meant to remove this extra logging.

Let me know if you have any questions!

### CODE REVIEW

**Comment by Vectorrent** [comment]
I noticed that the `Patcher` expects a static batch size to be set, but my custom model is more dynamic than that - and often uses a variable batch size. Thus, I need to adjust the batch size in the patcher, and made that possible with an override in the forward method.

**Comment by Vectorrent** [comment]
I reverted some of the `Patcher` code to it's original style, since I realized I could just use the `entropies` argument to handle the batch size stuff.

I also fixed a warning here, since passing `entropies` in that way was causing this:
```
bytelatent/data/patcher.py:535: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  scores = torch.tensor(entropies, dtype=torch.float32)
  ```

**Comment by Vectorrent** [comment]
Because of the way `calculate_entropies()` disables gradients, it is not possible to backpropagate through the weights of our entropy model, during training in realtime. The latest commit adds an optional way to enable those grads.

**Comment by Vectorrent** [comment]
In general, this seems to be working for me. I have been able to train an entropy model alongside of the latent model, simultaneously, with these changes. Let me know if you have any questions.

**Comment by lilitong7** [comment]
> This should be ready for review. I have been able to train an entropy model alongside of the latent model, simultaneously, with these changes. Let me know if you have any questions.

Hello, thank you very much for your suggestions and changes. The current code still does not provide a separate script for training the entropy_model, while you have provided code that supports loading both the BLT and entropy_model simultaneously. Could I use your code on top of the original train.py and run it? If so, are there any things I should pay attention to or modify? Thank you.

**Comment by Vectorrent** [comment]
@lilitong7, I am not using the original entropy model, nor any of the other components, really. I have completely replaced the encoder, decoder, latent model and entropy model with some other models, of my own design. Really, I'm just using all of the utility functions, to integrate BLT into my own project.

You're more than welcome to [look at my code](https://github.com/0-5788719150923125/praxis/blob/main/praxis/modules/encoder.py), which is fully-functional. But as for this repo, or any of the training scripts within - I haven't even tried to use them.

**Comment by lilitong7** [comment]
> @lilitong7, I am not using the original entropy model, nor any of the other components, really. I have completely replaced the encoder, decoder, latent model and entropy model with some other models, of my own design. Really, I'm just using all of the utility functions, to integrate BLT into my own project.
> 
> You're more than welcome to [look at my code](https://github.com/0-5788719150923125/praxis/blob/main/praxis/modules/encoder.py), which is fully-functional. But as for this repo, or any of the training scripts within - I haven't even tried to use them.
Thank you very much for your reply. I have carefully reviewed your code, and it seems you have used components such as the patcher and encoder from Byte Latent Transformer to build an improved version of the Byte Latent Transformer with an " encoder - long_model - decoder" structure, and you have also simulated the inference-stage functions. I’m wondering if you have trained the model using the same configurations as in the paper. I plan to do so, and if you have any results, could you share how they compare to the original model? Thanks again!

**Comment by Vectorrent** [comment]
@lilitong7 I have not trained the Byte Latent Transformer, using the same configurations from the paper. My architecture is quite different, and I have not yet trained it through to convergence (for lack of compute). 

So, the best I can say is: results look promising, based upon the metrics I have.
image

File: bytelatent/data/patcher.py [hunk 1]
```diff
@@ -24,8 +26,12 @@ class PatchingModeEnum(str, Enum):
 
 
 class PatcherArgs(BaseModel):
+    class Config:
+        arbitrary_types_allowed = True
+
     patching_mode: PatchingModeEnum = PatchingModeEnum.entropy
     patching_device: str = "cuda"
+    entropy_model: nn.Module | None = None
```
Comment (EntilZha) [inline bytelatent/data/patcher.py:34]: The configs here are suppose to be serializable and read from a file, so this would break this. I'm definitely open to having a way to set this more flexibly, but I don't think it should be in `PatcherArgs`.

I would instead suggest that you have a variable/flag here that is serializable, then create another function that can take those variables/flags to construct the `nn.Module` you supply here.

File: bytelatent/data/patcher.py [hunk 2]
```diff
@@ -58,7 +64,11 @@ def entropy(scores):
 
 
 def calculate_entropies(
-    tokens: torch.tensor, entropy_model, patching_batch_size, device: str | None = None
+    tokens: torch.tensor,
+    entropy_model,
+    patching_batch_size,
+    device: str | None = None,
+    enable_grad: bool = False,
```
Comment (EntilZha) [inline bytelatent/data/patcher.py:71]: Thanks for keeping default behavior the same

File: bytelatent/data/patcher.py [hunk 3]
```diff
@@ -101,6 +118,10 @@ def patch_start_mask_from_entropy_with_monotonicity(entropies, t):
     returns [bs, seq_len] mask where True indicates the start of a patch
     """
     bs, seq_len = entropies.shape
+
+    if seq_len == 0:
```
Comment (EntilZha) [inline bytelatent/data/patcher.py:122]: What case is this trying to catch?

File: bytelatent/data/patcher.py [hunk 4]
```diff
@@ -451,14 +476,19 @@ def __init__(self, patcher_args: PatcherArgs):
...
-            assert (
-                patcher_args.entropy_model_checkpoint_dir is not None
-            ), "Cannot require realtime patching without an entropy model checkpoint"
-            entropy_model = load_entropy_model(
-                patcher_args.entropy_model_checkpoint_dir
-            )
-            entropy_model, _ = to_device(entropy_model, patcher_args.patching_device)
-            self.entropy_model = entropy_model
+            if patcher_args.entropy_model is not None:
+                self.entropy_model = patcher_args.entropy_model
```
Comment (EntilZha) [inline bytelatent/data/patcher.py:480]: This is probably where I would add the bit of code that would construct the entropy model given the `patcher_args`. If this isn't possible with the caller that you have, then I think the other option is to have a flag to specifically not set the entropy model, then in the caller code set it like `patcher.entropy_model = my_model`.

File: bytelatent/model/local_models.py [hunk 5]
```diff
@@ -199,9 +199,6 @@ def init_weights(self, init_std=None):
 class LocalEncoder(LocalModelBase):
     def __init__(self, args: LocalModelArgs):
         super().__init__(args)
-        self.output_proj = (
```
Comment (EntilZha) [inline bytelatent/model/local_models.py:202]: Why remove this?

File: bytelatent/model/utils.py [hunk 6]
```diff
@@ -162,9 +162,6 @@ def create_causal_mask(
             return "causal"
 
         if BLT_SUPPRESS_ATTN_ERROR == 1:
-            logging.warning(
```
Comment (EntilZha) [inline bytelatent/model/utils.py:165]: I suppose I missed having this warn only once, I'm fine with deleting the warning and I can add it back in with a check for that, but feel free to add that in (would be a global var to check if the warning has been run once).

**Review by EntilZha — COMMENTED** [review]
Overall looks good to me, but it would be nice if `PatcherArgs` stays serializable, so avoiding having `entropy_model` be a `nn.Module`. I think there are two ways to get at this: (1) either pass the args that could be used to construct the entropy model or (2) pass an arg that will have the patcher set `patcher.entropy_model = None`, then the caller sets it specifically to the desired value.

File: bytelatent/model/utils.py (hunk 6 above)
Comment (Vectorrent) [reply bytelatent/model/utils.py:165]: The warning is still there. This was a duplicate of the exact same warning message... which already runs at initialization, and must be bypassed via the environment variable. [The warning still exists.](https://github.com/facebookresearch/blt/blob/e381743f6ee7b6edef86b3491dcef2c286ed0a83/bytelatent/model/utils.py#L166)

File: bytelatent/model/local_models.py (hunk 5 above)
Comment (Vectorrent) [reply bytelatent/model/local_models.py:202]: It doesn't actually do anything, and because `entropy_model_checkpoint_dir` doesn't exist on `LocalModelArgs` - it crashes when trying to use entropy patching. 

If this needs to exist, it should probably be as an assertion, rather than an unused variable.

File: bytelatent/data/patcher.py (hunk 3 above)
Comment (Vectorrent) [reply bytelatent/data/patcher.py:122]: When a sequence length is 1, these functions crash. I think it's because we remove the first token, earlier in the stack - making the sequence length 0 here.

It's an edge case most people won't run into, but I definitely did.

Comment (EntilZha) [reply bytelatent/data/patcher.py:118]: That makes sense.

File: bytelatent/model/local_models.py (hunk 5 above)
Comment (EntilZha) [reply bytelatent/model/local_models.py:202]: Thanks for the explanation, I missed the missing arg there so it makes sense to remove that.

File: bytelatent/model/utils.py (hunk 6 above)
Comment (EntilZha) [reply bytelatent/model/utils.py:165]: Makes sense, it probably is a bit verbose to fail and then also emit the warning so this is fine.

**Comment by EntilZha** [comment]
@Vectorrent this good to merge, at least as of now lgtm?

**Comment by Vectorrent** [comment]
> @Vectorrent this good to merge, at least as of now lgtm?

I've been using for a few days, and it works for me! I'd say go for it.

**Comment by EntilZha** [comment]
Great, thanks for the PR!
